### PR TITLE
Change device spec

### LINKED ISSUE #6394: Change device spec for numpy/cupy/intel64
Currently `chainer.get_device` and `chainer.using_device` accept python modules (`numpy`, `cupy`, `intel64`) as the device specifier. It's not necessarily intuitive.
After discussion, we decided to change as follows:

* For ChainerX backends, current spec is kept: specify as a string, e.g. `'native'`, `'native:0'`, `'cuda:1'`, etc.
* For other backends, prefix with a symbol `'@'`, e.g. `'@numpy'`, `'@cupy:1'`, `'@intel64'`.
* For convenience, integers are accepted for numpy/cupy backends (as both `str` and `int`):
  * `'-1'`, `-1` -> numpy
  * `'0'`, `0`, `'1'`, ... -> cupy

### PR DESCRIPTION
**Pull request by toslunar**
Close #6394

### CODE REVIEW

**Comment by toslunar** [comment]
In the PR I didn't add support for `'-1'`, `'0'`, `'1'`, ..., because I'd like to recommend the new spec for users who start using intel64 or chainerx backend.

Other choices for now:
- Tuple specs like `('native', 0)` are dropped.
- The functions keep accepting `chainerx.Device` and `chainer.backends.cuda.Device` objects.

**Comment by toslunar** [comment]
The error message will be improved by the PR.

Before:
```
>>> import chainer
>>> from chainer.backends import intel64
>>> chainer.get_device(intel64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tos/GitHub/chainer/chainer/backend.py", line 123, in get_device
    raise ValueError('Invalid device specifier: {}'.format(device_spec))
ValueError: Invalid device specifier: <module 'chainer.backends.intel64' from '/Users/tos/GitHub/chainer/chainer/backends/intel64.py'>
```

After:
```
>>> import chainer
>>> chainer.get_device('@intel64')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tos/GitHub/chainer/chainer/backend.py", line 140, in get_device
    return intel64.Intel64Device()
  File "/Users/tos/GitHub/chainer/chainer/backends/intel64.py", line 29, in __init__
    check_ideep_available()
  File "/Users/tos/GitHub/chainer/chainer/backends/intel64.py", line 103, in check_ideep_available
    'Reason: {}: {}'.format(type(_error).__name__, msg))
RuntimeError: iDeep is not available.
Reason: ModuleNotFoundError: No module named 'ideep4py'
```

**Comment by toslunar** [comment]
> In the PR I didn't add support for `'-1'`, `'0'`, `'1'`, ..., because I'd like to recommend the new spec for users who start using intel64 or chainerx backend.

I discussed with @asi1024, and decided to add support for `'-1'`, `'0'`, `'1'`, in order to run updated examples with the same command line arguments.

File: chainer/backend.py [hunk 1]
```diff
@@ -70,6 +71,15 @@ def copyto(dst, src):
             type(dst)))
 
 
+def _get_default_device_from_module(xp):
```
Comment (asi1024) [inline chainer/backend.py]: This function is used in another file. How about making it public?

Comment (toslunar) [reply chainer/backend.py]: Done.  I added comments because the function should not be used from a new interface.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
I found `get_device(numpy)` in examples.

**Comment by asi1024** [comment]
The tests seem to be passed, but how about refactoring other examples using the new `get_device`?

**Comment by toslunar** [comment]
I don't have time.

**Comment by toslunar** [comment]
```
% grep -- --gpu examples/**/*.py
examples/chainermn/dcgan/train_dcgan.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/chainermn/mnist/train_mnist.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/chainermn/mnist/train_mnist_checkpoint.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/chainermn/mnist/train_mnist_dual_parallel.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/chainermn/mnist/train_mnist_model_parallel.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/chainermn/parallel_convolution/train.py:    parser.add_argument('--gpu', '-g', action='store_true', default=False,
examples/chainermn/seq2seq/seq2seq.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/chainermn/seq2seq/seq2seq_mp1.py:    parser.add_argument('--gpu', '-g', action='store_true',
examples/cifar/train_cifar.py:    parser.add_argument('--gpu', '-g', type=int, default=0,
examples/cifar/train_cifar_custom_loop.py:    parser.add_argument('--gpu', '-g', type=int, default=0,
examples/dcgan/train_dcgan.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/image_captioning/predict.py:    parser.add_argument('--gpu', type=int, default=0,
examples/image_captioning/train.py:    parser.add_argument('--gpu', type=int, default=0,
examples/imagenet/train_imagenet.py:    group.add_argument('--gpu', '-g', type=int, nargs='?', const=0,
examples/imagenet/train_imagenet_data_parallel.py:    parser.add_argument('--gpus', '-g', type=int, nargs='*',
examples/memnn/test_memnn.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/memnn/train_memnn.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/mnist/inference.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/mnist/train_mnist.py:    group.add_argument('--gpu', '-g', type=int, nargs='?', const=0,
examples/mnist/train_mnist_custom_loop.py:    group.add_argument('--gpu', '-g', type=int, nargs='?', const=0,
examples/mnist/train_mnist_data_parallel.py:    parser.add_argument('--gpu0', '-g', type=int, default=0,
examples/mnist/train_mnist_data_parallel.py:    parser.add_argument('--gpu1', '-G', type=int, default=1,
examples/mnist/train_mnist_model_parallel.py:    parser.add_argument('--gpu0', '-g', default=0, type=int,
examples/mnist/train_mnist_model_parallel.py:    parser.add_argument('--gpu1', '-G', default=1, type=int,
examples/modelzoo/evaluate_caffe_net.py:parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/pix2pix/train_facade.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/pos/postagging.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/ptb/gentxt.py:    parser.add_argument('--gpu', type=int, default=-1,
examples/ptb/train_ptb.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/ptb/train_ptb_custom_loop.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/reinforcement_learning/ddpg_pendulum.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/reinforcement_learning/dqn_cartpole.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/sentiment/train_recursive_minibatch.py:    parser.add_argument('--gpu', '-g', default=-1, type=int,
examples/sentiment/train_sentiment.py:    parser.add_argument('--gpu', '-g', default=-1, type=int,
examples/seq2seq/seq2seq.py:    group.add_argument('--gpu', '-g', type=int, nargs='?', const=0,
examples/static_graph_optimizations/cifar/train_cifar.py:    parser.add_argument('--gpu', '-g', type=int, default=0,
examples/static_graph_optimizations/cifar/train_cifar_custom_loop.py:    parser.add_argument('--gpu', '-g', type=int, default=0,
examples/static_graph_optimizations/mnist/train_mnist.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/static_graph_optimizations/mnist/train_mnist_custom_loop.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/static_graph_optimizations/ptb/train_ptb_custom_loop.py:    parser.add_argument('--gpu', '-g', type=int, default=0,
examples/text_classification/run_text_classifier.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/text_classification/train_text_classifier.py:    parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/vae/train_vae.py:    parser.add_argument('--gpu', '-g', default=-1, type=int,
examples/wavenet/generate.py:parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/wavenet/train.py:parser.add_argument('--gpu', '-g', type=int, default=-1,
examples/word2vec/train_word2vec.py:    parser.add_argument('--gpu', '-g', default=-1, type=int,
```
```
% grep -- --device examples/**/*.py
examples/imagenet/train_imagenet.py:    parser.add_argument('--device', '-d', type=str, default='-1',
examples/imagenet/train_imagenet.py:                               'specify it with --device option.')
examples/mnist/train_mnist.py:    parser.add_argument('--device', '-d', type=str, default='-1',
examples/mnist/train_mnist_custom_loop.py:    parser.add_argument('--device', '-d', type=str, default='-1',
examples/seq2seq/seq2seq.py:    parser.add_argument('--device', '-d', type=str, default='-1',
```

**Comment by asi1024** [comment]
Jenkins, test this please.

File: chainer/backend.py [hunk 2]
```diff
@@ -70,6 +71,23 @@ def copyto(dst, src):
             type(dst)))
 
 
+def get_default_device_from_array_module(xp):
```
Comment (niboshi) [inline chainer/backend.py]: I think it's better to keep it private.
The term "default device" is not clearly defined. CUDA's current device is not "the default device".
There's also a discussion to introduce the concept of Chainer's "default device", which is different from this one.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
I opened #6661 for the remaining task.

File: examples/imagenet/train_imagenet.py [hunk 3]
```diff
@@ -72,20 +71,9 @@ def get_example(self, i):
 
 
 def parse_device(args):
```
Comment (niboshi) [inline examples/imagenet/train_imagenet.py]: I think this function should be removed.
You can simply write `device = chainer.get_device(args.gpu or args.device)`.

Comment (toslunar) [reply examples/imagenet/train_imagenet.py]: - `bool(0)` evaluates to `False` :cry:
- `args.device or args.gpu` looks OK, if it's fine to prioritize the `--device` arg.

Comment (niboshi) [reply examples/imagenet/train_imagenet.py]: I see... I don't think that works.
The default value of `--device` would have to be `''` (an empty string) which evaluates to `False`. That default value would not work by itself. As `--gpu` is deprecated, we should not rely on it as a fallback.

Another possibility is to write `device = chainer.get_device(args.device if args.gpu is None else args.gpu)`.

Comment (toslunar) [reply examples/imagenet/train_imagenet.py]: I'll try `dest='device'`.

File: examples/imagenet/train_imagenet.py [hunk 4]
```diff
@@ -140,22 +122,23 @@ def main():
...
     args = parser.parse_args()
 
-    device = parse_device(args)
-
     # Set the dtype if supplied.
     if args.dtype is not None:
         chainer.config.dtype = args.dtype
 
-    print('Device: {}'.format(device))
+    print('Device: {}'.format(args.device))
```
Comment (niboshi) [inline examples/imagenet/train_imagenet.py]: How about printing the normalized one?
The actual chosen device is not clear.

Comment (niboshi) [reply examples/imagenet/train_imagenet.py]: Thanks

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**